### PR TITLE
DEX-1667 combinations of assets in config

### PR DESCRIPTION
**Pull request by JennaWestenra**

### CODE REVIEW

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 1]
```diff
@@ -83,17 +83,56 @@ object OrderFeeSettings {
...
+      def filterCustomPairs(pred: AssetPair => Boolean): CustomAssetsSettings = copy(customPairs = customPairs.filter(pred))
+
+      def getSettings(assetPair: AssetPair): Option[OrderFeeSettings] =
+        settingsMap.get(assetPair)
+
+    }
+
+    object CustomAssetsSettings {
+
+      def apply(assets: Set[Asset], settings: OrderFeeSettings): CustomAssetsSettings = {
```
[the remark this answers is not in the document]
Comment (alexander-branevskiy) [reply dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:119]: consider this approach

**Review by alexander-branevskiy — CHANGES_REQUESTED** [review]

File: dex/src/main/scala/com/wavesplatform/dex/api/http/entities/HttpOrderFeeMode.scala [hunk 2]
```diff
@@ -125,16 +125,17 @@ object HttpOrderFeeMode {
...
-  }
+  def fromSettings(settings: OrderFeeSettings, matcherAccountFee: Long): HttpOrderFeeMode =
+    settings match {
+      case x: OrderFeeSettings.DynamicSettings => FeeModeDynamic(x.maxBaseFee + matcherAccountFee)
+      case OrderFeeSettings.FixedSettings(assetId, minFee) => FeeModeFixed(assetId, minFee)
+      case OrderFeeSettings.PercentSettings(assetType, minFee, minFeeInWaves) => FeeModePercent(assetType, minFee, minFeeInWaves)
+      case cs @ OrderFeeSettings.CompositeSettings(default, _, _, discount, _) =>
+        FeeModeComposite(
+          fromSettings(default, matcherAccountFee),
+          cs.getAllPairs.view.mapValues(fromSettings(_, matcherAccountFee)).toMap,
```
[the remark this answers is not in the document]
Comment (JennaWestenra) [reply dex/src/main/scala/com/wavesplatform/dex/api/http/entities/HttpOrderFeeMode.scala:136]: When we read fee config, we get lambda that requires function to validate asset pairs, so there will never be CompositeSettings with invalid pairs in CustomAssetsSettings.  Now it's useless to filter anything here

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 3]
```diff
@@ -83,17 +83,56 @@ object OrderFeeSettings {
...
+        val customPairs =
+          assets.foldLeft(Set.empty[AssetPair]) {
+            case (acc, elem) =>
+              acc ++ assets.map(asset => AssetPair(asset, elem))
+          }
+        CustomAssetsSettings(assets, settings, customPairs)
```
[the remark this answers is not in the document]
Comment (JennaWestenra) [reply dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala]: Useless, because now we make here a lambda with only valid pairs

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 4]
```diff
@@ -83,17 +101,64 @@ object OrderFeeSettings {
...
+
+      val customPairs =
+        assets.foldLeft(Set.empty[AssetPair]) {
+          case (acc, elem) =>
+            acc ++ assets.map(asset => AssetPair(asset, elem)).filter(validator)
+        }
+
+      lazy val settingsMap = customPairs.map(p => (p, settings)).toMap
+
+      def filterCustomPairs(pred: AssetPair => Boolean): CustomAssetsSettings = ???
```
Comment (stavegan) [inline dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:140]: ???

**Review by stavegan — CHANGES_REQUESTED** [review]

**Comment by stavegan** [comment]
Probably, a separate abstraction for calculating commissions based on configs should be created and tested individually

**Review by stavegan — APPROVED** [review]

File: dex/src/main/scala/com/wavesplatform/dex/model/ValidationStages.scala [hunk 5]
```diff
@@ -35,7 +35,8 @@ object ValidationStages {
     lastProcessedOffset: => Long,
     blacklistedAddresses: Set[Address],
     hasMatcherAccountScript: Boolean,
-    handleProofs: Order => Order
+    handleProofs: Order => Order,
+    pairValidator: AssetPair => Boolean
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/model/ValidationStages.scala:39]: no usage

File: dex/src/main/scala/com/wavesplatform/dex/settings/MatcherSettings.scala [hunk 6]
```diff
@@ -77,18 +82,45 @@ case class MatcherSettings(
...
+      }
 
 }
 
 object MatcherSettings extends ConfigReaders {
 
   implicit val byteStrConfigReader = byteStr58ConfigReader
 
-  implicit val longOrderFeeConfigReader = genericMapReader[Long, OrderFeeSettings] { x =>
+  type RawOrderFeeSettings = AssetPairValidator => OrderFeeSettings
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/settings/MatcherSettings.scala:97]: PartialOrderFeeSettings

File: dex/src/main/scala/com/wavesplatform/dex/settings/MatcherSettings.scala [hunk 7]
```diff
@@ -77,18 +82,45 @@ case class MatcherSettings(
...
+
+  implicit val orderFeeSettingsReader: ConfigReader[RawOrderFeeSettings] = ConfigReader.fromCursor[RawOrderFeeSettings] { cursor =>
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/settings/MatcherSettings.scala:99]: partialOrderFeeSettingsReader

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 8]
```diff
@@ -17,11 +18,16 @@ sealed trait OrderFeeSettings extends Product with Serializable
 
 object OrderFeeSettings {
 
+  type RawCustomAssetSettings = (AssetPair => Boolean) => CompositeSettings.CustomAssetsSettings
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:21]: ```type PartialCustomAssetSettings = AssetPairValidator => CompositeSettings.CustomAssetsSettings```

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 9]
```diff
@@ -17,11 +18,16 @@ sealed trait OrderFeeSettings extends Product with Serializable
...
+
+  final case class UnfilteredOrderFeeSettings(prepareValue: AssetPairValidator => OrderFeeSettings)
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:23]: there is no usage of the class

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 10]
```diff
@@ -83,17 +89,55 @@ object OrderFeeSettings {
...
     def getOrderFeeSettings(assetPair: AssetPair): OrderFeeSettings =
-      custom.getOrElse(assetPair, default)
+      custom.get(assetPair).orElse {
+        customAssets.flatMap(_.getSettings(assetPair))
+      }.getOrElse(default)
+
+    def getAllPairs: Map[AssetPair, OrderFeeSettings] =
+      customAssets.fold(Map.empty[AssetPair, OrderFeeSettings])(
+        _.settingsMap
+      ) ++ custom // custom must override any values from customAssets
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:105]: either remove comment or rephrase it to sth like:
```//custom has higher priority compared to customAssets```


there is a rule not commenting directly what code does =)
I've already seen that ```++``` will replace some assets from settingsMap

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 11]
```diff
@@ -83,17 +89,55 @@ object OrderFeeSettings {
...
 
   }
 
   object CompositeSettings extends ConfigReaders {
 
+    final case class CustomAssetsSettings(assets: Set[Asset], settings: OrderFeeSettings)(validator: AssetPairValidator) {
+
+      val customPairs =
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:113]: let's make this private

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 12]
```diff
@@ -83,17 +89,55 @@ object OrderFeeSettings {
...
+      lazy val settingsMap = customPairs.map(p => (p, settings)).toMap
+
+      def getSettings(assetPair: AssetPair): Option[OrderFeeSettings] =
+        settingsMap.get(assetPair)
+
+    }
+
+    object CustomAssetsSettings {
+
+      type RawCustomAssetsSettings = AssetPairValidator => CustomAssetsSettings
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:128]: Raw -> Partial 
and the same for all places below

File: dex/src/main/scala/com/wavesplatform/dex/tool/ConfigChecker.scala [hunk 13]
```diff
@@ -149,7 +150,11 @@ sealed trait ConfigWriters {
         )
     }
 
-  implicit val longOrderFeeConfigWriter: ConfigWriter[Map[Long, OrderFeeSettings]] = genericMapWriter(_.toString)
+  implicit val orderFeeSettingsWriter: ConfigWriter[RawOrderFeeSettings] =
+    orderFeeWriter
+      .contramap[RawOrderFeeSettings](uofs => uofs(_ => true))
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/tool/ConfigChecker.scala:155]: just ```(_(_ => true))```

**Review by alexander-branevskiy — CHANGES_REQUESTED** [review]

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 14]
```diff
@@ -83,17 +89,55 @@ object OrderFeeSettings {
...
     zeroFeeAccounts: Set[PublicKey] = Set.empty
   ) extends OrderFeeSettings {
 
     def getOrderFeeSettings(assetPair: AssetPair): OrderFeeSettings =
-      custom.getOrElse(assetPair, default)
+      custom.get(assetPair).orElse {
+        customAssets.flatMap(_.getSettings(assetPair))
+      }.getOrElse(default)
+
+    def getAllPairs: Map[AssetPair, OrderFeeSettings] =
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:102]: getAllCustomSettings

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 15]
```diff
@@ -83,17 +89,55 @@ object OrderFeeSettings {
...
+      ) ++ custom // custom must override any values from customAssets
 
   }
 
   object CompositeSettings extends ConfigReaders {
 
+    final case class CustomAssetsSettings(assets: Set[Asset], settings: OrderFeeSettings)(validator: AssetPairValidator) {
+
+      val customPairs =
+        assets.foldLeft(Set.empty[AssetPair]) {
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:114]: it's better do filter one time 
```
assets.foldLeft(Set.empty[AssetPair]) {
          case (acc, elem) =>
            acc ++ assets.map(asset => AssetPair(asset, elem))
        }.filter(validator)
```

File: dex/src/main/scala/com/wavesplatform/dex/tool/ConfigChecker.scala (hunk 13 above)
Comment (stavegan) [reply dex/src/main/scala/com/wavesplatform/dex/tool/ConfigChecker.scala:155]: looks awful

File: dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala [hunk 16]
```diff
@@ -83,17 +89,55 @@ object OrderFeeSettings {
...
+    def getAllPairs: Map[AssetPair, OrderFeeSettings] =
+      customAssets.fold(Map.empty[AssetPair, OrderFeeSettings])(
+        _.settingsMap
+      ) ++ custom // custom must override any values from customAssets
 
   }
 
   object CompositeSettings extends ConfigReaders {
 
+    final case class CustomAssetsSettings(assets: Set[Asset], settings: OrderFeeSettings)(validator: AssetPairValidator) {
```
Comment (alexander-branevskiy) [inline dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:111]: let's avoid curring in case classes. just use
```
final case class CustomAssetsSettings(assets: Set[Asset], settings: OrderFeeSettings, validator: AssetPairValidator)
```

Comment (alexander-branevskiy) [reply dex/src/main/scala/com/wavesplatform/dex/settings/OrderFeeSettings.scala:111]: and CustomAssetsSettings(assets, settings, _) for lambda variant

File: dex/src/test/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRouteSpec.scala [hunk 17]
```diff
@@ -134,6 +134,32 @@ class MatcherApiRouteSpec extends RouteSpec("/matcher") with MatcherSpecBase wit
...
+
+  private val simpleCompositeSettings = CompositeSettings(
+    default = DynamicSettings(baseMakerFee = 350000, baseTakerFee = 350000),
+    custom = Map(
+      assetPair1 -> PercentSettings(AssetType.Amount, minFee = 0.01, minFeeInWaves = 1000),
+      assetPair2 -> PercentSettings(AssetType.Amount, minFee = 0.02, minFeeInWaves = 1000)
+    )
+  )
+
+  private val customCompositeSettings = CompositeSettings(
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRouteSpec.scala:151]: complexCompositeSettings

**Review by alexander-branevskiy — APPROVED** [review]
